ipydatagrid/datagrid.py

```diff
@@ -982,3 +983,196 @@ def _set_renderer_defaults(self):
+                    ]
+                elif operator == "startswith":
+                    dataframe = dataframe[
+                        dataframe[column].str.startswith(value)
+                    ]
+                elif operator == "endswith":
+                    dataframe = dataframe[dataframe[column].str.endswith(value)]
+                elif operator == "stringContains":
+                    dataframe = dataframe[dataframe[column].str.contains(value)]
+                elif operator == "contains":
+                    dataframe = dataframe[dataframe[column].str.contains(value)]
+                elif operator == "!contains":
+                    dataframe = dataframe[
+                        not dataframe[column].str.contains(value)
+                    ]
+                elif operator == "isOnSameDay":
+                    value = pd.to_datetime(value).date()
+                    dataframe = dataframe[
+                        pd.to_datetime(dataframe[column]).dt.date == value
+                    ]
+                else:
+                    raise RuntimeError(
+                        f"Unrecognised filter operator '{operator}'"
+                    )
+
+        return dataframe
+
+    def _handle_comm_msg(self, _, content, _buffs):
+        event_type = content.get("type", "")
+
+        if event_type == "data-request":
+            r1 = content.get("r1")
+            r2 = content.get("r2")
+            c1 = content.get("c1")
+            c2 = content.get("c2")
+
+            # Filter/sort whole dataset before selecting rows/cols of interest
+            if self._transformed_data is not None:
+                # Use existing transformed data.
+                value = self._transformed_data
+            else:
+                value = self.__dataframe_reference
+
+            value = value.iloc[r1 : r2 + 1, c1 : c2 + 1]
+
+            # Primary key used
+            index_key = self.get_dataframe_index(value)
+
+            serialized = _data_serialization_impl(
+                self.generate_data_object(value, "ipydguuid", index_key), None
+            )
+
+            # Extract all buffers
+            buffers = []
+            for column in serialized["data"].keys():
+                if (
+                    not isinstance(serialized["data"][column], list)
+                    and not serialized["data"][column]["type"] == "raw"
+                ):
+                    buffers.append(serialized["data"][column]["value"])
+                    serialized["data"][column]["value"] = len(buffers) - 1
+
+            answer = {
+                "event_type": "data-reply",
+                "value": serialized,
+                "r1": r1,
+                "r2": r2,
+                "c1": c1,
+                "c2": c2,
+            }
+
+            self.send(answer, buffers)
+
+        elif event_type == "frontend-transforms":
+            # Transforms is an array of dicts.
+            frontend_transforms = content.get("transforms")
```

Comment:
Instead of going through a custom comm messages for transforms requests, we should have the front-end change the `_transforms` property in the widget model and have the Python side handle the change on this property.

```suggestion
    @observe("_transforms")
    def _on_transforms_changed(self, change):
        # Transforms is an array of dicts.
        frontend_transforms = change["new"]
```

This will also help us having the same code being triggered whether transforms are applied from the UI or from the Python API (through the `transform` method)